Add scaled_mm python API, test

slayton58 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.12.0) (oldest at bottom):
* __->__ #164142
* #164141

Summary:

* Add `torch.nn.functional.scaled_mm` as an abstraction around the C++
  methods
* Wraps `torch._scaled_mm_v2` API by default, but user can force use of
  the older `torch._scaled_mm` interface.
* Scaled MM tests now run on the new API

Test Plan:

`pytest test/test_scaled_matmul_cuda.py`

Reviewers:

Subscribers:

Tasks:

Tags:
Signed-off-by: Simon Layton <simonlaytonmeta.com>

slayton58 committed:
Update

[ghstack-poisoned]

slayton58 commented:
@pytorchbot rebase

slayton58 committed:
Update

[ghstack-poisoned]

slayton58 committed:
Update

[ghstack-poisoned]

slayton58 committed:
Update

[ghstack-poisoned]

slayton58 committed:
Update

[ghstack-poisoned]

slayton58 committed:
Update

[ghstack-poisoned]

slayton58 committed:
Update

[ghstack-poisoned]

slayton58 committed:
Update

[ghstack-poisoned]

slayton58 committed:
Update

[ghstack-poisoned]

slayton58 commented on docs/source/quantization.rst:
```diff
@@ -3,6 +3,8 @@
 Quantization
 ============
 
+.. automodule:: torch.quantization.scaled_mm
```
Post some offline discussion, moved to `torch.nn.functional.scaled_mm`

slayton58 committed:
Update

[ghstack-poisoned]

slayton58 committed:
Update

[ghstack-poisoned]

slayton58 committed:
Update

[ghstack-poisoned]

slayton58 committed:
Update

[ghstack-poisoned]

slayton58 committed:
Update

[ghstack-poisoned]

slayton58 committed:
Update

[ghstack-poisoned]

slayton58 committed:
Update

[ghstack-poisoned]

slayton58 committed:
Update

[ghstack-poisoned]

slayton58 committed:
Update

[ghstack-poisoned]

slayton58 committed:
Update

[ghstack-poisoned]

slayton58 committed:
Update

[ghstack-poisoned]

slayton58 commented on test/test_fx.py:
```diff
@@ -4657,6 +4657,7 @@ def tearDown(self):
         "adaptive_avg_pool3d": LEN_ERROR,
         "adaptive_max_pool2d_with_indices": LEN_ERROR,
         "adaptive_max_pool3d_with_indices": LEN_ERROR,
+        "scaled_mm": LEN_ERROR,
```
`len(...)` is non-traceable by default (errors out during the test), and `scaled_mm` uses `len` for some list processing - this prevents the test from running.

slayton58 replied:
I guess so - I see 2 immediate ways to fix this:
* remove the error-checking for the deprecated fallback path and just pass the first scale etc. from the passed list, and rely on erroring out in C++ with an invalid scaling recipe
* Remove the deprecated fallback path entirely

drisspg commented on torch/nn/functional.py:
```diff
@@ -6509,3 +6509,121 @@ def multi_head_attention_forward(
+    if len(kwargs) > 0:
+        raise RuntimeError("kwargs contains unexpected entries, ", kwargs.keys())
+
+    if use_deprecated_api:
```
OOC why this?

slayton58 committed:
Update

[ghstack-poisoned]

slayton58 committed:
Update

[ghstack-poisoned]

slayton58 committed:
Update

[ghstack-poisoned]

slayton58 committed:
Update

[ghstack-poisoned]

slayton58 committed:
Update

[ghstack-poisoned]

slayton58 committed:
Update on "Add scaled_mm python API, test"


Summary:

* Add `torch.nn.functional.scaled_mm` as an abstraction around the C++
  methods
* Wraps `torch._scaled_mm_v2` API by default, but user can force use of
  the older `torch._scaled_mm` interface.
* Scaled MM tests now run on the new API

Test Plan:

`pytest test/test_scaled_matmul_cuda.py`

Reviewers:

Subscribers:

Tasks:

Tags:
Signed-off-by: Simon Layton <simonlaytonmeta.com>


[ghstack-poisoned]

slayton58 committed:
Update on "Add scaled_mm python API, test"


Summary:

* Add `torch.nn.functional.scaled_mm` as an abstraction around the C++
  methods
* Wraps `torch._scaled_mm_v2` API by default, but user can force use of
  the older `torch._scaled_mm` interface.
* Scaled MM tests now run on the new API

Test Plan:

`pytest test/test_scaled_matmul_cuda.py`

Reviewers:

Subscribers:

Tasks:

Tags:
Signed-off-by: Simon Layton <simonlaytonmeta.com>


[ghstack-poisoned]

drisspg reviewed (approved):
Looks good in a follow up we should have some more robust composability testing, the immediate things that come to mind is writing meta_registrations.py entry adding some compile tests and then making sure we (for now) fallback in inductor and then ultimately rewire the lowerings.

cc @eellison  is the best way to get most of this testing still through common_methods_invocations.py?

slayton58 committed:
Update on "Add scaled_mm python API, test"


Summary:

* Add `torch.nn.functional.scaled_mm` as an abstraction around the C++
  methods
* Wraps `torch._scaled_mm_v2` API by default, but user can force use of
  the older `torch._scaled_mm` interface.
* Scaled MM tests now run on the new API

Test Plan:

`pytest test/test_scaled_matmul_cuda.py`

Reviewers:

Subscribers:

Tasks:

Tags:
Signed-off-by: Simon Layton <simonlaytonmeta.com>


[ghstack-poisoned]

slayton58 committed:
Update on "Add scaled_mm python API, test"


Summary:

* Add `torch.nn.functional.scaled_mm` as an abstraction around the C++
  methods
* Wraps `torch._scaled_mm_v2` API by default, but user can force use of
  the older `torch._scaled_mm` interface.
* Scaled MM tests now run on the new API

Test Plan:

`pytest test/test_scaled_matmul_cuda.py`

Reviewers:

Subscribers:

Tasks:

Tags:
Signed-off-by: Simon Layton <simonlaytonmeta.com>


[ghstack-poisoned]

slayton58 committed:
Update

[ghstack-poisoned]
